pkg/health/health.go

```diff
@@ -11,89 +11,75 @@ import (
 	"github.com/sirupsen/logrus"
 )
 
-// DiagnosisResult is a possible result after a diagnosis.
-type DiagnosisResult int
+// Diagnosis is a possible result after a diagnosis.
+type Diagnosis int
 
 // Possible diagnosis results.
 const (
-	Unknown DiagnosisResult = iota
+	Unknown Diagnosis = iota
 	Inconclusive
 	Healthy
 	Unhealthy
 )
 
-// Diagnosis is the information about the health of the revision.
-type Diagnosis struct {
-	OverallResult DiagnosisResult
-	CheckResults  []CheckResult
-}
-
-// CheckResult is information about a metrics criteria check.
-type CheckResult struct {
-	Threshold     float64
-	ActualValue   float64
-	IsCriteriaMet bool
-}
-
 // Diagnose attempts to determine the health of a revision.
 //
+// If no health criteria is specified or the size of the health criteria and the
+// actual values are not the same, the diagnosis is Unknown and an error is
+// returned.
+//
 // If the minimum number of requests is not met, then health cannot be
-// determined and Diagnosis.EnoughRequests is set to false.
+// determined and diagnosis is Inconclusive.
 //
-// Otherwise, all metrics criteria are checked to determine if the revision is
-// healthy.
-func Diagnose(ctx context.Context, provider metrics.Provider, query metrics.Query,
-	offset time.Duration, minRequests int64, healthCriteria []config.Metric) (*Diagnosis, error) {
-
+// Otherwise, all metrics criteria are checked to determine whether the revision
+// is healthy or not.
+func Diagnose(ctx context.Context, healthCriteria []config.Metric, actualValues []float64) (Diagnosis, error) {
 	logger := util.LoggerFromContext(ctx)
-	metricsValues, err := CollectMetrics(ctx, provider, query, offset, healthCriteria)
-	if err != nil {
-		return nil, errors.Wrap(err, "could not collect metrics")
+	if len(healthCriteria) != len(actualValues) {
+		return Unknown, errors.New("the size of health criteria is not the same to the size of the actual metrics values")
+	}
+	if len(healthCriteria) == 0 {
+		return Unknown, errors.New("health criteria must be specified")
 	}
 
-	overallResult := Healthy
-	var results []CheckResult
-	for i, criteria := range healthCriteria {
+	diagnosis := Healthy
```

Comment:
Still not handling `Inconclusive`. It will be added later, see #35

Reply:
Consider this approach.
https://play.golang.org/p/ikdSjj1F0ae

Here I define a bitmap and mark if I've seen a "inconclusive" or "unhealthy". At the end I make a decision.

Reply:
This fails when we remove the `Inconclusive` result https://play.golang.org/p/IcOr8bjz_4L
The problem is that `Unhealthy` gets 6. If we want the bitmap to work, we need the values to be [1, 2, 4, 8] or [0, 2, 4, 8] instead of [0, 2, 4, 6]
I think it was just a typo in your example since this works https://play.golang.org/p/yym1xTQD9y-

I'm still a little worried about readability though.

Reply:
Ok. Thinking about it. If we look at the conditionals at https://play.golang.org/p/yym1xTQD9y-, we can see that priorities are set. That is, we care about results in the following order
1) Are there any inconclusive results?
2) Are there any unmet criteria?
3) Are all criteria met?
4) Unknown

Notice that unknown can only happen if no 1-3 do not, but we are already checking that you have at least a metrics criteria, so 1-3 must happen if you get to that point. 

We can also just start with Unknown and set the health to `Healthy` only the first time a criteria is met:
```
// Criteria is met at this point
if diagnosis == Unknown {
   diagnosis = Healthy
}
```

We will have a special case for request count anyways, so we can break at that point if no enough requests were made and return `Inconclusive`

With this approach, we're basically maintaining the same priority as with the bitmap, and in my view, we make it simpler to read.

Reply:
OK sounds good. We don't have to initialize as Unknown either if you think the code is error-proof for future iterations (potentially by others).

Reply:
Still want to initialize as `Unknown` just in case the only health criteria is request count.